File: LeetCode 75 Study Plan to Ace Interviews/Q205 - Isomorphic Strings/solution.cs
public class Solution {
    public bool IsIsomorphic(string s, string t) {
        // solution 1: using hash map by Dictionary
        if (s.Length != t.Length) {
            return false;
        }
        Dictionary<char, char> charMap = new Dictionary<char, char>();
        for (int i = 0; i < s.Length; i++) {
            if (charMap.ContainsKey(s[i])) {
                if(t[i] != charMap[s[i]]){
                    return false;  
                }
            } else if (charMap.ContainsValue(t[i])) {
                return false;
            } else {
                charMap.Add(s[i], t[i]);
            }     
        }
        return true;
    }
}

File: LeetCode 75 Study Plan to Ace Interviews/Q205 - Isomorphic Strings/solution.py
# solution 1 - hash map by dict: 40ms, 15.3MB
class Solution1:
    def isIsomorphic(self, s: str, t: str) -> bool:

        if len(s) != len(t):
            return False
        charMap = {}
        for index in range(len(s)):
            if s[index] in charMap:
                if (t[index] != charMap.get(s[index])):
                    return False
            elif t[index] in charMap.values():
                return False
            else:
                charMap[s[index]] = t[index]
        return True


# solution 2 - set and zip: 32ms, 15.4MB
# comparing the length of unique char in s and t
# as well as the length of unique pair in dict paired by s, t
class Solution2:
    def isIsomorphic(self, s: str, t: str) -> bool:
        numOfUniqueCharsInS = len(set(s))
        numOfUniqueCharsInT = len(set(t))
        stDict = zip(s, t)
        numOfUniqueCharsInStDict = len(set(zip(s, t)))
        return numOfUniqueCharsInS == numOfUniqueCharsInT == numOfUniqueCharsInStDict
